src/Materials/Textures/Loaders/envTextureLoader.ts

```diff
@@ -42,15 +42,21 @@ export class _ENVTextureLoader implements IInternalTextureLoader {
             texture.width = info.width;
             texture.height = info.width;
 
-            EnvironmentTextureTools.UploadEnvSpherical(texture, info);
-            EnvironmentTextureTools.UploadEnvLevelsAsync(texture, data, info).then(() => {
-                texture.isReady = true;
-                texture.onLoadedObservable.notifyObservers(texture);
-                texture.onLoadedObservable.clear();
-                if (onLoad) {
-                    onLoad();
-                }
-            });
+            try {
+                EnvironmentTextureTools.UploadEnvSpherical(texture, info);
+                EnvironmentTextureTools.UploadEnvLevelsAsync(texture, data, info).then(() => {
+                    texture.isReady = true;
+                    texture.onLoadedObservable.notifyObservers(texture);
+                    texture.onLoadedObservable.clear();
+                    if (onLoad) {
+                        onLoad();
+                    }
+                }).catch((reason) => {

```

Comment:
nit: it's a tiny bit cheaper to do this

```suggestion
                }, (reason) => {
```

Reply:
Looking at it a bit more I actually think the `.catch()` approach here is what we want. It does add an extra layer of promise resolution but with that approach it will also catch if the .then() throws an error inside of the onSuccess callback. Which in this case is quite plausible since it is triggering the onLoadedObservable callbacks.

Happy to change it though if you think it should still use the passed in onReject rather than the catch chaining.

Reply:
Interesting. I think the code probably shouldn't call both onLoad and onError. With the way you had it, if onLoad threw an exception, then it would call onError. The way I suggested doesn't quite work either since it won't trap the error coming [lines 48-50](from https://github.com/BabylonJS/Babylon.js/pull/9835/files/cdd5d02666605eb4a681d60042d1cd8aeca1d947#diff-e7a1c1509ff5bf5a669adbcb9718d819b466e40b0916e7bfa9e8b5b95810a70aR48-R50).

We probably need something like this:
```javascript
                EnvironmentTextureTools.UploadEnvLevelsAsync(texture, data, info).then(() => {
                    texture.isReady = true;
                    texture.onLoadedObservable.notifyObservers(texture);
                    texture.onLoadedObservable.clear();
                }.then(() =>{
                    if (onLoad) {
                        onLoad();
                    }
                }, reason) => {
                    onError?.("Can not upload environment levels", reason);
                });
```

Reply:
Wait, this isn't right either. We shouldn't throw for the onLoadedObservable either. So what I suggested is right I think.

Reply:
One thing that is a bit weird currently is if the `onLoad` callback throws, then the `onError` callback is also called. It might make sense to do something more like:
```ts
EnvironmentTextureTools.UploadEnvLevelsAsync(texture, data, info).then(() => {
    texture.isReady = true;
    texture.onLoadedObservable.notifyObservers(texture);
    texture.onLoadedObservable.clear();
}).then(onLoad, (reason) => onError?.("Can not upload environment levels", reason));
```

Reply:
That is more or less my suggestion although originally for the wrong reason.

Reply:
Going back to the original suggestion so that it's up to the onLoad callbacks to handle any errors they might generate themselves.